File: husky_mpc/husky_mpc/robot.py
import numpy as np
import casadi as ca


class Robot:
	def __init__(self, x, y, omega, safety_radius, linear_v_min, linear_v_max, angular_v_min,  angular_v_max, dt):
		
		self.safety_radius = safety_radius

		self.linear_v_min = linear_v_min 
		self.linear_v_max = linear_v_max
		self.angular_v_min = angular_v_min
		self.angular_v_max = angular_v_max

		self.location = np.array([
							[x],
							[y],
							[omega]
						  ])

		self.x_dot = np.array([
							[0],
							[0],
							[0]
						  ])

		self.wheel_speed = np.array([
										[0],
										[0]
									])

		self.b = 0.25
		self.r = 0.05

		self.car_dims = np.array([
										[-self.b, -self.b, 1],
										[0 		, -self.b, 1],
										[ self.b,  		0, 1],
										[ 0, 	   self.b, 1],
										[ -self.b, self.b, 1]
									])
		
		self.B_matrix = np.array([
						[np.cos(self.location[2, 0])*dt,  0],
						[np.sin(self.location[2, 0])*dt,  0],
						[0					 , dt]
					])
		



	# Computes wheel velocities with linear and angular velocities
	def inverse_kinematics(self):
		ikine_mat = np.array([
							[1/self.r, 0, self.b/self.r],
							[1/self.r, 0, -self.b/self.r]
							])

		return ikine_mat@self.x_dot
	

	# Step 1 (after MPC): Update Robot Volocity
	def set_robot_velocity(self, linear_velocity, angular_velocity):
		self.x_dot = np.array([
										[linear_velocity],
										[0],
										[angular_velocity]
									])
		# # self.wheel_speed = self.inverse_kinematics()






	# Step 2 (after MPC): Move Robot
	def update(self, dt):
		# # self.wheel_speed[self.wheel_speed>2] = 2;
		# # self.wheel_speed[self.wheel_speed<-2] = -2;

		# recompute linear and angular velocities using adjusted wheel speeds
		# # self.x_dot = self.forward_kinematics() 
		
		self.update_state(dt)
		# self.wheel_speed = self.inverse_kinematics()


	def forward_kinematics(self):
		kine_mat = np.array([
							[self.r/2  		  , self.r/2],
							[0 		 		  ,	0],
							[self.r/(2*self.b), -self.r/(2*self.b)]
							])

		return kine_mat@self.wheel_speed


	def update_state(self, dt):
		A = np.array([
						[1, 0, 0],
						[0, 1, 0],
						[0, 0, 1]
					])
		B = np.array([
						[np.cos(self.location[2, 0])*dt,  0],
						[np.sin(self.location[2, 0])*dt,  0],
						[0					 , dt]
					])
		

		vel = np.array([
							[self.x_dot[0, 0]],
							[self.x_dot[2, 0]]
						])

		self.location = A@self.location + B@vel


	def get_state(self):
		return self.location, self.x_dot

	

File: husky_mpc_cpp/src/Robot.cpp
#include "husky_mpc_cpp/Robot.hpp"
#include <iostream>
#include <Eigen/Dense>

Robot::Robot(double linear_v, double angular_v, double x, double y, double omega,
    double linear_v_max, double angular_v_max, double dt): 
        current_linear_v_(linear_v), current_angular_v_(angular_v), 
        linear_v_max_(linear_v_max), angular_v_max_(angular_v_max)
{
    //Eigen::VectorXd location_ = Eigen::VectorXd({x, y, omega});

    // x_dot = Eigen::VectorXd({0, 0, 0});
    // wheel_speed_ = Eigen::VectorXd({0, 0});

    // b_ = 0.25;
    // r_ = 0.05;

    // car_dims = Eigen::MatrixXd({{-b_, -b_, 1},
    //                             {0, -b_, 1},
    //                             {b_, 0, 1},
    //                             {0, b_, 1},
    //                             {-b_, b_, 1}});
    

    /*
    Eigen::MatrixXd B_matrix_(3,2);
    B_matrix_(0,0) = cos(location_(2, 0))*dt;
    B_matrix_(0,1) = 0;
    B_matrix_(1,0) = sin(location_(2, 0))*dt;
    B_matrix_(1,1) = 0;
    B_matrix_(2,0) = 0;
    B_matrix_(2,1) = dt;
    */

    // ikine_mat_ = Eigen::MatrixXd({{1/r_, 0, b_/r_},
    //                             {1/r_, 0, -b_/r_}});

    // kine_mat_ = Eigen::MatrixXd({{r_/2, r_/2},
    //                             {0, 0},
    //                             {r_/(2*b_), -r_/(2*b_)}});
}

// Eigen::VectorXd Robot::inverse_kinematics()
// {
//     return ikine_mat_ * x_dot_;
// }

// void Robot::set_robot_velocity(double linear_velocity, double angular_velocity)
// {
//     x_dot_(0) = linear_velocity;
//     x_dot_(1) = 0;
//     x_dot_(2) = angular_velocity;
// }

// void Robot::update_state(double dt)
// {
//     Eigen::MatrixXd A = Eigen::MatrixXd::Identity(3, 3);
//     Eigen::MatrixXd B = Eigen::MatrixXd({{cos(location_(2, 0))*dt, 0},
//                                         {sin(location_(2, 0))*dt, 0},
//                                         {0, dt}});

//     Eigen::VectorXd vel = Eigen::VectorXd({x_dot_(0), x_dot_(2)});
//     location_ = A*location_ + B*vel;
// }

// Eigen::VectorXd Robot::forward_kinematics()
// {
//     return kine_mat_ * wheel_speed_;
// }